### DISCUSSION TITLE
Can array be both ref observable and deep observable at the same time?

### DISCUSSION THREAD

**Discussion by Lechros** [mobxjs/mobx#3262] [Q&A] [answered]
Suppose I have SomeClass[] array. I want some code to react to change deeply, changes in SomeClass should notify the code. I want another code to only observe changes in the array itself, array.push and assigns should update, but changes to SomeClass should be ignored. Both codes observe the same array.
Is it possible to do this? Is it my best option to observe deeply and manually set logic inside reaction to distinguish two cases?

**Comment by urugator** [comment]
You can disable subscriptions with `untracked`
```javascript
autorun(() => {
  array.forEach(item => {
    untracked(() => {
      console.log(item) // item changes ignored
    })
  })
})
```

**Comment by mweststrate** [comment, marked as answer]
To add some more context, please check out https://mobx.js.org/understanding-reactivity.html, what changes reactions depend on doesn't depend on what is observable (that is just the prerequisite), but on what is read. So if one reaction deeply reads the data, and the other one only shallowly the array, one will respond to deep changes and the other one to array changes only. There is nothing special that needs to be done to achieve that.

When you do read some data in a reaction, but still don't want to react to them in the future, then `untracked` comes into play as suggested by @urugator, but these cases are in practice pretty rare.

**Reply by Lechros** [reply]
Thanks I think I now understand how reaction works.
I'm trying to save objects inside array or Map to storage. It would be best to re-save only changed object, how can I get changed index/key/value? Current code example:
```
autorun(() => {
  for(const [key, data] of datas) {
    localforage.setItem(key, data)
  }
)
```
This would re-save all items when one of them changes. I want to get notified of added, removed key / changed value('s key).

**Reply by urugator** [reply]
Take a look at [`observe`](https://mobx.js.org/intercept-and-observe.html#observe), but note it works completely differently than `reaction/autorun`.
You probably have some actions that manipulate the array, so I would suggest to simply move the persistence logic there.

**Reply by Lechros** [reply]
Maybe I was too obsessed with mobx's "reactive" thingy. Adding persistence logic to action seems good, I should probably do that. Thanks!
